File: src/org/genyris/core/DynamicSymbol.java
package org.genyris.core;

import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;
import org.genyris.interp.UnboundException;

public class DynamicSymbol extends Symbol {

	private Symbol _realSymbol;
	
	public DynamicSymbol(Symbol sym) {
		_realSymbol = sym;
	}
    public int compareTo(Object arg0) {
		return ((DynamicSymbol) arg0)._realSymbol.compareTo(_realSymbol);
    }

    public boolean equals(Object arg0) {
    	if(!(arg0 instanceof DynamicSymbol)) {
    		return false;
    	}
    	return _realSymbol.equals(((DynamicSymbol)arg0)._realSymbol);
    }

    public int hashCode() {
        return _realSymbol.hashCode();
    }

    public void acceptVisitor(Visitor guest) throws GenyrisException {
		guest.visitDynamicSymbol(this);
	}

	public Symbol getBuiltinClassSymbol(Internable table) {
		return table.DYNAMICSYMBOLREF();
	}
    public String toString() {
        return "."+ _realSymbol.toString();
    }

	public String getPrintName() {
		return "."+ _realSymbol.getPrintName();
	}
	public Symbol getRealSymbol() {
		return _realSymbol;
	}
	public void defineVariable(Environment env, Exp valu) throws GenyrisException {
		env.defineDynamicVariable(this, valu);
	}
	public Exp lookupVariableValue(Environment env) throws UnboundException {
		return env.lookupDynamicVariableValue(this);
	}
	public void setVariableValue(Environment env, Exp valu) throws UnboundException {
		env.setDynamicVariableValue(this, valu);
	}

}

File: src/org/genyris/core/DynamicSymbol.kt
package org.genyris.core

import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.interp.UnboundException

class DynamicSymbol(val realSymbol: Symbol) : Symbol() {
    override fun compareTo(arg0: Any?): Int {
        if (arg0 !is DynamicSymbol) {
            return -1
        }
        return arg0.realSymbol.compareTo(this.realSymbol)
    }

    override fun equals(arg0: Any?): Boolean {
        if (arg0 !is DynamicSymbol) {
            return false
        }
        return this.realSymbol == arg0.realSymbol
    }

    override fun hashCode(): Int {
        return realSymbol.hashCode()
    }

    @kotlin.Throws(GenyrisException::class)
    override fun acceptVisitor(guest: Visitor) {
        guest.visitDynamicSymbol(this)
    }

    override fun getBuiltinClassSymbol(table: Internable): Symbol? {
        return table.DYNAMICSYMBOLREF()
    }

    override fun toString(): String {
        return "." + realSymbol.toString()
    }

    override fun getPrintName(): String {
        return "." + realSymbol.getPrintName()
    }

    @kotlin.Throws(GenyrisException::class)
    override fun defineVariable(env: Environment, valu: Exp?) {
        env.defineDynamicVariable(this, valu)
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupVariableValue(env: Environment): Exp? {
        return env.lookupDynamicVariableValue(this)
    }

    @kotlin.Throws(UnboundException::class)
    override fun setVariableValue(env: Environment, valu: Exp?) {
        env.setDynamicVariableValue(this, valu)
    }
}
